src/SourceLocator/Ast/Parser/MemoizingParser.php

```diff
@@ -4,27 +4,42 @@
 
 namespace Roave\BetterReflection\SourceLocator\Ast\Parser;
 
+use Closure;
+use JsonSerializable;
+use PhpParser\Comment;
 use PhpParser\ErrorHandler;
+use PhpParser\JsonDecoder;
+use PhpParser\Node;
+use PhpParser\NodeAbstract;
 use PhpParser\Parser;
 use function array_key_exists;
+use function array_map;
 use function hash;
-use function serialize;
+use function is_array;
 use function strlen;
-use function unserialize;
 
 /**
  * @internal
  */
 final class MemoizingParser implements Parser
 {
-    /** @var string[] indexed by source hash */
+    /** @var array<string, Node\Stmt[]> indexed by source hash */
     private array $sourceHashToAst = [];
 
     private Parser $wrappedParser;
 
+    private JsonDecoder $jsonDecoder;
+
+    private Closure $decode;
+
     public function __construct(Parser $wrappedParser)
     {
         $this->wrappedParser = $wrappedParser;
+        $this->jsonDecoder   = new JsonDecoder();
+        $this->decode        = Closure::bind(function (array $value) {
+            /** @psalm-suppress UndefinedMethod */
+            return $this->decodeRecursive($value);
```

Comment:
Hmm, no, we really shouldn't do this.

This is `private` API of `PhpParser` that we shouldn't touch at all: doing a `json_encode()` before passing a structure to the decoder is preferable, even if it comes with a performance impact.

Alternatively, we port `JsonDecoder` (with tests) completely, but overall we could do something much simpler:

 1. deprecate `MemoizingParser` and leave it as-is
 2. create a new version of `MemoizingParser` that keeps AST nodes in-memory as they are (no cloning nor anything), but we do mention that explicitly in naming and documentation somehow (that it isn't safe to change the AST)

Reply:
Understood :) I can remove the last commit. The first two improved the memory too.

Reply:
Done.